Improve pruned package data loading and handling

dsplaisted opened the pull request:
Improves how [package pruning data](https://github.com/NuGet/Home/blob/dev/accepted/2024/prune-package-reference.md) is supplied in the .NET SDK.

- Prune any later patches of a pruned package (by setting the patch value to 32767).  See https://github.com/dotnet/sdk/issues/44566
- Correctly supply pruned package data for .NET Framework, .NET Standard, and framework references using profiles (such as WPF and Windows Forms)
- Use pruned package data from targeting packs for .NET 10 and higher.

Targeting packs include a PackageOverrides.txt which provides the data we can use to determine the packages that should be pruned.  However, this data was not entirely accurate for previous versions of .NET, as it was meant as an optimization so it didn't matter much if a few packages were missing or the versions listed weren't up to date.  As of .NET 10, that should be fixed, so we will use the PackageOverrides.txt data as the source for package pruning for .NET 10 and higher.

For earlier versions of .NET, and for .NET Standard and .NET Framework, we will use data that has been calculated based on running conflict resolution and curated to be correct.  This data is encapsulated in the FrameworkPackages source files.

During the .NET SDK build, targeting packs are downloaded for previous versions of .NET that are version 10.0 or higher, and the PackageOverrides.txt files from those targeting packs are laid out in a PrunePackageData folder in the SDK layout.  Since there aren't currently any "previous versions of .NET that are version 10.0 or higher", this won't actually do anything until .NET 11.

The GetPackagesToPrune task will load data from the FrameworkPackages, the PrunePackageData folder, or the targeting packs for the current version of .NET that are bundled with the SDK, depending on the target framework and version.

dsplaisted committed:
Include package prune data in SDK layout

dsplaisted committed:
Refactor package prune data cache code

dsplaisted committed:
Generate Package Prune Data in stage 2 redist

dsplaisted committed:
Rename task

dsplaisted committed:
Compare prune package data from framework packages and PrunePackageData

dsplaisted committed:
Change source for prune package data depending on target framework

dsplaisted committed:
Remove code to compare Prune Package Data and framework packages data

dsplaisted committed:
Add prune package tests for different target frameworks, and fix loading packages for them

dsplaisted committed:
Map framework reference profiles when getting pruned packages

dsplaisted committed:
Remove comment

dsplaisted committed:
Prune any patches of pruned packages

ViktorHofer commented:
One important use-case are repositories that build shared frameworks (targeting packs) and consume them in their live build. These repositories need to be able to tell the SDK to use their live built targeting pack PackageOverrides.txt file. The current use of `NetCoreTargetingPackRoot` seems to make this hard. Such repos only build one targeting pack, i.e. `Microsoft.NETCore.App.Ref`. Therefore overriding the property wouldn't be correct as it would also impact the resolution of the pruned package data of the other non live-built targeting packs.

FWIW runtime uses this hook to tell the SDK to use the locally built packs: https://github.com/dotnet/runtime/blob/0004083b20a30dbb030aeb7e977feca0187c856c/eng/targetingpacks.targets#L119-L132

ViktorHofer commented:
> For repositories that build shared frameworks, is it important to prune packages? Do those repos have transitive references to packages that should be pruned?

Yes, let me speak for runtime but this should also apply to aspnetcore and windowsdesktop. Runtime builds the live shared framework and then uses that to build the out-of-band package projects and tests. OOBs and test projects have external dependencies that bring transitives in that overlap with the framework assemblies. Those already caused CG issues in the past and it would be good to be able to use package pruning here.

> Is this problem specific to package overrides, or is it a general problem for using a repo-built targeting pack for one shared framework but also referencing another shared framework? How is that handled today?

This is specific to the PackageOverrides.txt file. The SDK offers all the hooks needed to point to live built packs. I.e. the PlatformManifest.txt and other data files are used from the live built SDK.

ericstj commented:
I'll add that we want our repos to use the product in the best way possible -- that way our builds test what we ship to customers.  The more custom we need to make our solution the more we have to maintain and we lose valuable coverage of the SDK features.

nkolev92 commented on src/Layout/redist/targets/GeneratePackagePruneData.targets:
```diff
@@ -0,0 +1,60 @@
+      <Output TaskParameter="PackageDownloads" ItemName="TargetingPackForPruneDataCollated" />
+    </CollatePackageDownloads>
+
+    <!-- Can't run custom tasks in the build before NuGet restore, so create a separate project and restore that to download the packages -->
```
When is this run?

dsplaisted replied:
This is run during the build of the SDK repo.  We need to run restore in order to build the tasks which are used in building the repo, so we can't run custom tasks before NuGet restore.  So this moves the PackageDownload of the targeting packs to a separate project which we restore separately.

nkolev92 replied:
Ah cool. 
The targeting pack solution, but at SDK build time.

dsplaisted committed:
Test fixes

dsplaisted committed:
Bring back CollatePackageDownloads task for package pruning

This partially reverts commit c0bfcdbb90a6aec6c562039b8f00eb0fd4dc202f.

# Conflicts:
#	src/Installer/redist-installer/targets/BundledTemplates.targets

dsplaisted commented:
> artifacts/obj/extracted-dotnet-sdk/sdk/10.0.100-ci/Containers/build/Microsoft.NET.Build.Containers.targets(6,9): error MSB4184: The expression "[MSBuild]::VersionGreaterThan('', 7.0.100)" cannot be evaluated. Version string was not in a correct format. [/__w/1/vmr/src/scenario-tests/artifacts/obj/generatedtests/SdkTemplateTestsComplex_Console_CSharp/SdkTemplateTestsComplex_Console_CSharp.csproj] [/__w/1/vmr/src/scenario-tests/src/Microsoft.DotNet.ScenarioTests.SdkTemplateTests/Microsoft.DotNet.ScenarioTests.SdkTemplateTests.csproj]

@dotnet/product-construction @baronfel Can someone take a look at this VMR build issue?  Is it a known issue showing up elsewhere?  Thanks.

MiYanni commented on src/Tasks/Microsoft.NET.Build.Tasks/GetPackagesToPrune.cs:
```diff
@@ -72,51 +80,180 @@ protected override void ExecuteCore()
+                }
+            }
+
+            bool useFrameworkPackageData;
+            if (!key.TargetFrameworkIdentifier.Equals(".NETCoreApp") || targetFrameworkVersion.Major < 10)
+            {
+                //  Use hard-coded / generated "framework package data" for .NET 9 and lower, .NET Framework, and .NET Standard
+                useFrameworkPackageData = true;
+            }
+            else
+            {
+                //  Use bundled "prune package data" for .NET 10 and higher.  During the redist build, this comes from targeting packs and
+                //  is laid out in the PrunePackageData folder.
+                useFrameworkPackageData = false;
+            }
```
nit:

```suggestion
            //  When true, use hard-coded / generated "framework package data" for .NET 9 and lower, .NET Framework, and .NET Standard
            // When false, use bundled "prune package data" for .NET 10 and higher.  During the redist build, this comes from targeting packs and is laid out in the PrunePackageData folder.
            bool useFrameworkPackageData = !key.TargetFrameworkIdentifier.Equals(".NETCoreApp") || targetFrameworkVersion.Major < 10;
```

MiYanni commented on src/Tasks/Microsoft.NET.Build.Tasks/GetPackagesToPrune.cs:
```diff
@@ -72,51 +80,180 @@ protected override void ExecuteCore()
+        {
+            foreach (var package in packagesToAdd)
+            {
+                if (packagesToPrune.TryGetValue(package.id, out NuGetVersion existingVersion))
+                {
+                    if (package.version > existingVersion)
+                    {
+                        //  There are some "inconsistent" versions in the FrameworkPackages data.  This happens because, for example, the ASP.NET Core shared framework for .NET 9 inherits
+                        //  from the ASP.NET Core shared framework for .NET 8, but not from the base Microsoft.NETCore.App framework for .NET 9.  So for something like System.IO.Pipelines,
+                        //  which was in ASP.NET in .NET 8 but moved to the base shared framework in .NET 9, we will see an 8.0 version from the ASP.NET shared framework and a 9.0 version
+                        //  from the base shared framework.  As long as the base shared framework is always referenced together with the ASP.NET shared framework, this shouldn't be a
+                        //  problem, and we can just pick the latest version of the package that we see.
+                        packagesToPrune[package.id] = package.version;
+                    }
+                }
+                else
+                {
+                    packagesToPrune[package.id] = package.version;
+                }
```
nit:

```suggestion
                // There are some "inconsistent" versions in the FrameworkPackages data.  This happens because, for example, the ASP.NET Core shared framework for .NET 9 inherits
                // from the ASP.NET Core shared framework for .NET 8, but not from the base Microsoft.NETCore.App framework for .NET 9.  So for something like System.IO.Pipelines,
                // which was in ASP.NET in .NET 8 but moved to the base shared framework in .NET 9, we will see an 8.0 version from the ASP.NET shared framework and a 9.0 version
                // from the base shared framework.  As long as the base shared framework is always referenced together with the ASP.NET shared framework, this shouldn't be a
                // problem, and we can just pick the latest version of the package that we see.
                if (!packagesToPrune.TryGetValue(package.id, out NuGetVersion existingVersion) || package.version > existingVersion)
                {
                    packagesToPrune[package.id] = package.version;
                }
```

richlander commented:
Someone once said to me that "the SDK cannot build the SDK" and I did not initially understand what was meant by that simple but provocative statement. It has stuck with me.

dsplaisted commented:
> > artifacts/obj/extracted-dotnet-sdk/sdk/10.0.100-ci/Containers/build/Microsoft.NET.Build.Containers.targets(6,9): error MSB4184: The expression "[MSBuild]::VersionGreaterThan('', 7.0.100)" cannot be evaluated. Version string was not in a correct format. [/__w/1/vmr/src/scenario-tests/artifacts/obj/generatedtests/SdkTemplateTestsComplex_Console_CSharp/SdkTemplateTestsComplex_Console_CSharp.csproj] [/__w/1/vmr/src/scenario-tests/src/Microsoft.DotNet.ScenarioTests.SdkTemplateTests/Microsoft.DotNet.ScenarioTests.SdkTemplateTests.csproj]
> 
> @dotnet/product-construction @baronfel Can someone take a look at this VMR build issue? Is it a known issue showing up elsewhere? Thanks.

@dotnet/product-construction @dotnet/source-build Ping on this, can someone take a look at why the source build / VMR builds here are failing?

mthalman commented:
> Can someone take a look at this VMR build issue? Is it a known issue showing up elsewhere? Thanks.

It's not a known issue. Somehow these changes are causing the `NetCoreSdkVersion` property to be empty in the SDK produced with these changes. It's probably worthwhile to grab the built SDK from the pipeline artifacts and examine it (Windows_x64_Artifacts_Attempt1/assets/Release/Sdk).

dsplaisted committed:
Apply suggestions from code review

Co-authored-by: Michael Yanni <MiYanni@microsoft.com>

dsplaisted committed:
Code review feedback and comment cleanup

dsplaisted committed:
Fix copy/paste issue in GenerateBundledVersions.targets

dsplaisted committed:
Support overriding or providing additional paths for prune package data

This will support repos that are building a targeting pack and then using it in their build

dsplaisted commented:
> > Can someone take a look at this VMR build issue? Is it a known issue showing up elsewhere? Thanks.
> 
> It's not a known issue. Somehow these changes are causing the `NetCoreSdkVersion` property to be empty in the SDK produced with these changes. It's probably worthwhile to grab the built SDK from the pipeline artifacts and examine it (Windows_x64_Artifacts_Attempt1/assets/Release/Sdk).

Thanks, it turned out this was my mistake but it was in the generation of the bundled versions file which currently isn't applied for our normal test runs.

dsplaisted commented:
@ViktorHofer @ericstj I've updated this so that you can set a `PrunePackageTargetingPackRoots` property, which can be a semicolon-separated list of where targeting packs can be found.

Would this work for you?  It essentially allows you to specify multiple targeting pack roots just for the purposes of looking up the PackageOverrides.txt file to get the prune package data.  You'll still have to have the full folder layout, so under the root you'd need the package contents to be under something like `Microsoft.NETCore.App.Ref\<PackageVersion>`.

dsplaisted committed:
Don't try to get package prune data for frameworks that don't have it.

Also fix some tests

dsplaisted committed:
Fix empty framework reference in prune package data

dsplaisted commented:
@mthalman @dotnet/product-construction This is failing for the VMR build in the aspnetcore build.repotasks build.  It's failing in new code I'm adding in this PR, and it looks like it's failing because the targeting packs aren't included in the layout of the SDK it's using.

Do we expect the targeting packs to be in the packs directory for the SDK that's being used here?

Is there a way to see the source build layout?  I don't seem to see it in the build artifacts.

dsplaisted commented:
> @dsplaisted I assume we can't depend on the `ResolveFrameworkReferences` target to get that information? Asking as that's the hook that we use in runtime and SBRP to set the local paths: https://github.com/dotnet/runtime/blob/4c6a7068e9024696cb0bb0f996b89cb0f03adb32/eng/targetingpacks.targets#L120-L122
> 
> That target probably doesn't run during restore but I still wanted to ask.

Yes, `ResolveFrameworkReferences` needs the targeting packs to be downloaded, so it can only run after restore.  `GetPackagesToPrune` has to run before restore because it generates the pruned packages list which is needed for restore.  So `GetPackagesToPrune` has to duplicate some of the logic that `ResolveFrameworkReferences` has.

dsplaisted committed:
Add logging for GetPackagesToPrune

dsplaisted commented:
> The build that's failing is the stage 2 source build leg. In that same pipeline, the leg that's passing is stage 1. In stage 1, we use the Microsoft-built SDK to build the VMR in source build mode. In stage 2, we take the SDK produced by stage 1 and build the VMR again using that SDK. This verifies that the source-built SDK can be used to build the VMR. So the issue seems to be with how the SDK layout is produced by stage 1. You can find that SDK in the pipeline artifacts: `CentOSStream9_Online_MsftSdk_x64_Artifacts/assets/Release/dotnet-sdk-<version>-centos.9-x64.tar.gz`

@mthalman Thanks, that's helpful.  I looked in the stage 1 layout, and there is a `packs\Microsoft.NETCore.App.Ref\10.0.0-preview.3.25154.11` folder in it.  However, in the stage 2 build it doesn't seem to be finding that folder.  I added some logging to the GetPackagesToPrune task:

```
Looking for targeting packs in /__w/1/vmr/src/aspnetcore/.dotnet/packs/Microsoft.NETCore.App.Ref
Targeting pack folder /__w/1/vmr/src/aspnetcore/.dotnet/packs/Microsoft.NETCore.App.Ref does not exist
```

Should the `/__w/1/vmr/src/aspnetcore/.dotnet/` folder have the same contents as the stage 1 layout from the .tar.gz file?

dsplaisted committed:
Add error message if prune package data can't be found, and add property to allow missing data

dsplaisted committed:
Add localization comment

ViktorHofer commented:
`/__w/1/vmr/src/aspnetcore/.dotnet/` shouldn't exist. When building inside the VMR, the SDK is restored once in the VMR root and used to build the repositories. That would `/__w/1/vmr/.dotnet/` in this example.

Sounds like something incorrectly points to the wrong location.

ViktorHofer commented on src/Layout/redist/targets/GeneratePackagePruneData.targets:
```diff
@@ -0,0 +1,60 @@
+      <Output TaskParameter="PackageDownloads" ItemName="TargetingPackForPruneDataCollated" />
+    </CollatePackageDownloads>
+
+    <!-- Can't run custom tasks in the build before NuGet restore, so create a separate project and restore that to download the packages -->
```
This worries me. I submitted various PRs in the last month that removed these in-built restores. That has multiple downsides:
- The build doesn't work without network connection after an initial `build.cmd/sh -restore`. "Air plane mode"
- With a coming SFI requirement, the times when we can hit the network will be restricted.
- Increased complexity and the build is slower because of the intermediate projects.

We hit that same limitation in SBRP (https://github.com/dotnet/source-build-reference-packages/blob/232bcf31aad21949f80d6706720540b85e43fff3/src/packageSourceGenerator/PackageSourceGenerator.proj#L98C9-L104) and decided to introduce an inline task that can run before restore. Now that I see this use-case here as well, I would prefer to move the `CollatePackageDownloads` task into the Arcade SDK so that the task is available before project restore.

dsplaisted replied:
It's not just the collating, there's also the `GeneratePackagePruneDataDownloads` task.

It would be nice if we can find a good pattern for running custom tasks that help determine what needs to be restored.

ViktorHofer replied:
We can follow-up on this post-merge. We should avoid these in-build downloads. My guess is that we won't be able to do that anymore with a future SFI wave.

ViktorHofer commented:
> @ViktorHofer @ericstj I've updated this so that you can set a PrunePackageTargetingPackRoots property, which can be a semicolon-separated list of where targeting packs can be found.
Would this work for you? It essentially allows you to specify multiple targeting pack roots just for the purposes of looking up the PackageOverrides.txt file to get the prune package data. You'll still have to have the full folder layout, so under the root you'd need the package contents to be under something like Microsoft.NETCore.App.Ref\<PackageVersion>.

Yes that will work to repos that build live targeting pakcs. Thanks for responding to my feedback. Just curious, why is that a property instead of an item?

ViktorHofer commented on src/Tasks/sdk-tasks/GeneratePackagePruneDataDownloads.cs:
```diff
@@ -0,0 +1,58 @@
+            //  HOWEVER: we don't download the targeting pack for the maximum .NET version here, as we may still be in preview.
+            //  Rather, the GetPackagesToPrune task will load the package prune data for the current version from the
+            //  targeting packs that ship with the SDK.
+            for (int netVersion = 10; netVersion < maxNetVersion; netVersion++)
```
Isn't this defacto a no-op today with .NET 10 as `netVersion` is equal to `maxNetVersion`?

ViktorHofer replied:
OK you mentioned this in the PR description. Resolving.

MiYanni reviewed (approved):
I just looked over the files that changed since the last time I reviewed. Looks good!

dsplaisted commented:
> `/__w/1/vmr/src/aspnetcore/.dotnet/` shouldn't exist. When building inside the VMR, the SDK is restored once in the VMR root and then used to build _all_ the repositories. That would `/__w/1/vmr/.dotnet/` in this example.
> 
> Something points to the wrong location. What is that?

@ViktorHofer Thanks!  The aspnetcore repo is overriding the `NetCoreTargetingPackRoot` property: https://github.com/dotnet/aspnetcore/blob/f3555640d3b0d049856947c4f2bd0b869adf5c5e/Directory.Build.props#L193-L204

I think this is probably related to the issue you brought up earlier in this PR - that some repos build their own ref packs and need to be able to use the locally-built ref packs and package overrides for the framework they built while using the previous stage's ref packs and package overrides for other frameworks.

I think what we should do for now is to set the `AllowMissingPrunePackageData` property to true for the aspnetcore build in order to unblock this PR.  Once it's in then someone from ASP.NET or the VMR team or you (not sure) can work on using the `PrunePackageTargetingPackRoots` in the aspnetcore build in order to enable correct package pruning during its build.

Does that make sense?  Do we need to add a VMR patch to this PR to set the `AllowMissingPrunePackageData` for the aspnetcore build?

ViktorHofer commented:
> Thanks! The aspnetcore repo is overriding the NetCoreTargetingPackRoot property:

I don't think that's it as that property is conditioned on `'$(DotNetBuild) != 'true'` but the failure is happening when building aspnetcore inside the VMR (for which `DotNetBuild` is true). The binlog in the pipeline artifacts should provide more information.

ViktorHofer commented:
OK looks like this Exec is not passing all the required properties through: https://github.com/dotnet/aspnetcore/blob/f3555640d3b0d049856947c4f2bd0b869adf5c5e/eng/DotNetBuild.props#L48

@wtgodbe @mmitche I think the `DotNetBuildOrchestrator` and `DotNetBuildInnerRepo` properties are missing here which makes `DotNetBuild` false: https://github.com/dotnet/arcade/blob/6c1093b3501b0b08f4db5817f1d633a0f59267b2/src/Microsoft.DotNet.Arcade.Sdk/tools/RepoDefaults.props#L17

wtgodbe commented:
So we'd need to just pass `InnerBuildArgs` to the Exec tasks?

dsplaisted committed:
Merge upstream changes

# Conflicts:
#	src/Layout/redist/redist.csproj

ViktorHofer commented on src/Layout/redist/redist.csproj:
```diff
@@ -18,6 +18,7 @@
     <!-- The D.B.targets file is in subfolder so that the tools projects don't import it. -->
     <DirectoryBuildTargetsPath>$(MSBuildThisFileDirectory)targets\Directory.Build.targets</DirectoryBuildTargetsPath>
   </PropertyGroup>
+  <Import Project="targets\GeneratePackagePruneData.targets" />
```
This import should be moved into targets\Directory.Build.targets.

ViktorHofer commented:
My change that just got merged created new conflicts: https://github.com/dotnet/sdk/commit/ad154d6a67abc2d884e0f75495770e33b18d7b0c

I will push a commit to resolve them. Hope that's OK.

ViktorHofer committed:
Merge branch 'main' into prune-package-data

dsplaisted commented:
> My change that just got merged created new conflicts: [ad154d6](https://github.com/dotnet/sdk/commit/ad154d6a67abc2d884e0f75495770e33b18d7b0c)
> 
> I will push a commit to resolve them. Hope that's OK.

@ViktorHofer Thanks a bunch, I really appreciate this!  Both the merge of redist and redist-installer as well as the fixup to my PR.

Now it looks like there's a new source build / VMR test failing: Microsoft.DotNet.Tests.JoinVerticalsTests.JoinVerticalsCheckAllAmbiguitiesTest.  Can someone take a look at that? @dotnet/product-construction

ViktorHofer commented:
Of course, it was a pleasure consolidating that infra :)

For this PR, I think we still need to wait for https://github.com/dotnet/aspnetcore/commit/63fd3321e0254c497e6b03cf0767a30151747062 to flow into sdk.

> Now it looks like there's a new source build / VMR test failing: Microsoft.DotNet.Tests.JoinVerticalsTests.JoinVerticalsCheckAllAmbiguitiesTest. Can someone take a look at that?

That should go away when you merge with latest main.

dsplaisted committed:
Merge upstream changes

ViktorHofer committed:
Merge branch 'main' into prune-package-data

ViktorHofer reviewed (approved):
I didn't closely review the implementation but overall LGTM. @ericstj would you mind also taking a peek at this?

ericstj commented on src/Tasks/Microsoft.NET.Build.Tasks/GetPackagesToPrune.cs:
```diff
@@ -72,51 +83,208 @@ protected override void ExecuteCore()
+        {
+            var nugetFramework = new NuGetFramework(targetFrameworkIdentifier, Version.Parse(targetFrameworkVersion));
+
+            //  FrameworkPackages just has data for .NET Framework 4.6.1 and doesn't handle framework compatibility, so treat anything greater than .NET Framework as if it were 4.6.1
```
I think this does handle compatibility - https://github.com/dotnet/sdk/blob/b6fef1d9dc68e04ff93587328dfd45a707874943/src/Tasks/Microsoft.NET.Build.Tasks/FrameworkPackages/FrameworkPackages.cs#L97

dsplaisted replied:
I disabled that in most cases because I want it to error out if it can't find a framework except in specific cases.  But I've changed this logic to re-enable the compatibility for .NET Framework instead of resetting the TargetFramework it's looking up.

ericstj commented on src/Tasks/Microsoft.NET.Build.Tasks/GetPackagesToPrune.cs:
```diff
@@ -72,51 +83,208 @@ protected override void ExecuteCore()
+                //  See https://github.com/dotnet/sdk/issues/44566
+                //  To do this, we set the patch version for the package to be pruned to 32767, which should be
+                //  higher than any actual patch version.
+                var maxPatch = new NuGetVersion(p.Value.Major, p.Value.Minor, 32767);
```
I don't think we should do this for .NETFramework - since there we version the assembly on every package release.

Will this apply for .NETFramework?

dsplaisted committed:
Use framework fallback for .NET Framework 4.6.1 or higher

dsplaisted committed:
Don't prune latest patch versions except for .NET Core

ericstj commented on src/Tasks/Microsoft.NET.Build.Tasks/GetPackagesToPrune.cs:
```diff
@@ -72,51 +83,219 @@ protected override void ExecuteCore()
-                item.SetMetadata("Version", p.Value.ToString());
+
+                string version;
+                if (key.TargetFrameworkIdentifier.Equals(".NETCoreApp", StringComparison.OrdinalIgnoreCase))
```
Was having some discussion with @ViktorHofer about this and it might also be better to avoid this lifting when `!p.Value.IsPrerelease`.   What are your thoughts on this @dsplaisted?
```suggestion
                if (key.TargetFrameworkIdentifier.Equals(".NETCoreApp", StringComparison.OrdinalIgnoreCase) && !p.Value.IsPrerelease)
```

dsplaisted committed:
Apply code review feedback

Co-authored-by: Eric StJohn <ericstj@microsoft.com>

dsplaisted committed:
Fix test

dsplaisted commented:
@wtgodbe @ViktorHofer Now that this is merged, I believe that it will cause the aspnetcore build to start failing once this SDK flows to the stage 0 for that repo.  The repo will need to be updated to set `AllowMissingPrunePackageData` to true, or (ideally) to lay out the targeting pack correctly and set `PrunePackageTargetingPackRoots` so that the prune package data for the just-built targeting pack can be used.

wtgodbe commented:
> or (ideally) to lay out the targeting pack correctly and set PrunePackageTargetingPackRoots so that the prune package data for the just-built targeting pack can be used.

@dsplaisted Could you elaborate on what we need to do to get it laid out currently? Currently we use the Arcade `PublishToDisk` to lay it out to a folder in `artifacts/obj`: https://github.com/dotnet/aspnetcore/blob/a182f3558f229cd4fa7a781e7b2e9fbed94aaad5/src/Framework/App.Ref/src/Microsoft.AspNetCore.App.Ref.sfxproj#L185-L189

dsplaisted commented:
> > or (ideally) to lay out the targeting pack correctly and set PrunePackageTargetingPackRoots so that the prune package data for the just-built targeting pack can be used.
> 
> @dsplaisted Could you elaborate on what we need to do to get it laid out currently? Currently we use the Arcade `PublishToDisk` to lay it out to a folder in `artifacts/obj`: https://github.com/dotnet/aspnetcore/blob/a182f3558f229cd4fa7a781e7b2e9fbed94aaad5/src/Framework/App.Ref/src/Microsoft.AspNetCore.App.Ref.sfxproj#L185-L189

You can set `PrunePackageTargetingPackRoots` to a list of folders where the SDK will look for the PackageOverrides.txt files.  Each folder needs to be laid out the same as the `dotnet/packs` folder, so you would need a folder that has subfolders for `Microsoft.NETCore.App.Ref\<PackageVersion>`, and under that would have the same layout as the targeting pack contents.

So you may need to move the output around so it's in the right layout.  Or, looking at the target you linked, it's possible it's already in the right folder and in the right layout, so it might be that there aren't any changes necessary.
